primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -75,10 +77,133 @@ where
 
     /// Enforce correct `merkle_proof` for the `elem` against
     /// `expected_merkle_root`.
-    fn enforce_merkle_proof(
+    fn enforce_membership_proof(
         &mut self,
         elem: Self::LeafVar,
         merkle_proof: Self::MerklePathVar,
         expected_merkle_root: Variable,
     ) -> Result<(), CircuitError>;
 }
+
+/// Gadget for the sparse merkle tree
+pub trait SparseMerkleTreeGadget<M>: MerkleTreeGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    /// checking non-membership proof
+    fn is_non_member(
+        &mut self,
+        elem: Self::LeafVar,
+        merkle_proof: Self::MerklePathVar,
+        merkle_root: Variable,
+    ) -> Result<BoolVar, CircuitError>;
+
+    /// Enforce correct `merkle_proof` for the empty leaf `elem` against
+    /// `expected_merkle_root`.
+    fn enforce_non_membership_proof(
```

Comment:
I think it might be better to have a consistent naming, by calling this `enforce_non_member`.

Reply:
I like `enforce_non_membership_proof` better, it's clear what it's enforcing.

Reply:
True. I agree that this is more clear.